test/Npgsql.Tests/SystemTransactionTests.cs

```diff
@@ -686,18 +707,22 @@ public static void EnlistVolatile(bool shouldRollBack = false, ConcurrentQueue<T
             public static void EscalateToDistributed(ConcurrentQueue<TransactionEvent> eventQueue)
                 => EscalateToDistributed(false, eventQueue);
 
-            public static void EscalateToDistributed(bool shouldRollBack= false, ConcurrentQueue<TransactionEvent> eventQueue = null)
+            public static void EscalateToDistributed(bool shouldRollBack = false, ConcurrentQueue<TransactionEvent> eventQueue = null)
                 => Enlist(true, shouldRollBack, eventQueue);
 
             static void Enlist(bool durable, bool shouldRollBack, ConcurrentQueue<TransactionEvent> eventQueue)
             {
-                var name = $"{(durable ? "Durable" : "Volatile")} resource";
+                Counter++;
+
+                var name = $"{(durable ? "Durable" : "Volatile")} resource {Counter}";
                 var resource = new EnlistResource(shouldRollBack, name, eventQueue);
                 if (durable)
                     Transaction.Current.EnlistDurable(Guid.NewGuid(), resource, EnlistmentOptions.None);
                 else
                     Transaction.Current.EnlistVolatile(resource, EnlistmentOptions.None);
 
+                Transaction.Current.TransactionCompleted += resource.Current_TransactionCompleted;
```

Comment:
Just to check where this one is inserted in the 2PC. It appears, anywhere after the first phase: can occurs before second phase, concurrently to it, or after, varying from one test run to the other. Sometimes it is even on the same thread than a second phase (obviously not occurring concurrently those times).